### PR TITLE
Feature/144 network cleaning simplifying functions

### PR DESCRIPTION
**Pull request by ArdtK**

### CODE REVIEW

File: ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py [hunk 1]
```diff
@@ -93,3 +96,151 @@ def from_shapefile_to_poly(shapefile: Path, out_path: Path, outname: str = ""):
...
+
+    # convert lat,lng (y,x) node to x,y for shapely distance operation
+    xy_point = Point(node_coor)
+
+    # calculate Euclidean distance from each edge's geometry to this node
+    edge_distances = [(edge, xy_point.distance(edge[4])) for edge in edges]
+
+    # the nearest edge minimizes the non-zero distance to the node
+    (u, v, data, key, geom), dist = min(edge_distances, key=lambda x: x[1] if x[1] > 0 else float('inf'))
+    utils.log(f"Found nearest edge ({u, v, data, key}) to node {node}")
```
Comment (Carsopre) [inline ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py:125]: do not log with a third party library, instead do:
```python
import logging
...
logging.info(f"Found nearest edge...")
```

File: ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py [hunk 2]
```diff
@@ -93,3 +96,151 @@ def from_shapefile_to_poly(shapefile: Path, out_path: Path, outname: str = ""):
...
+                    "nearest_edge": (u, v, data, key, dist)}
+        elif return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, key, geom)}
+        else:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, key)}
+
+
+def _is_endnode_simplified(graph: MultiDiGraph, node: int) -> bool:
```
Comment (Carsopre) [inline ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py:184]: This method should not be with  underscore `_`. This symbol represents that the method is private and should only be accessed by methods within the same context, however you are importing it from another module. 

Please do not forget to describe the docstrings.

File: ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py [hunk 3]
```diff
@@ -93,3 +96,151 @@ def from_shapefile_to_poly(shapefile: Path, out_path: Path, outname: str = ""):
...
+    if node in neighbors:
+        # if the node appears in its list of neighbors, it self-loops
+        # this is always an endpoint.
+        return True
+
+    # rule 2
+    elif graph.out_degree(node) == 0 or graph.in_degree(node) == 0:
+        # if node has no incoming edges or no outgoing edges, it is an endpoint
+        return True
+
```
Comment (Carsopre) [inline ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py:202]: What do you do if none of the above conditions is met? Do not forget to add a default return statement.

File: ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py [hunk 4]
```diff
@@ -93,3 +96,151 @@ def from_shapefile_to_poly(shapefile: Path, out_path: Path, outname: str = ""):
...
+        graph.add_node(new_node_id, x=new_node.x, y=new_node.y, geometry=new_node, **new_node_data)
+    else:
+        new_node_id = find_existing_node(graph, new_node)[0]
+    # Add the two new edges if do not exist already
+    graph = create_edge(graph, u=u, v=new_node_id)
+    graph = create_edge(graph, u=new_node_id, v=u)
+    # Remove the original edge
+    graph.remove_edge(u, v)
+
+def remove_key(element_data: dict, keys_to_exclude: list):
```
Comment (Carsopre) [inline ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py:222]: Output type hints + docstrings please.

File: ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py [hunk 5]
```diff
@@ -93,3 +96,151 @@ def from_shapefile_to_poly(shapefile: Path, out_path: Path, outname: str = ""):
...
+    # Remove geometry information from the the new_node_data
+    for key in keys_to_exclude:
+        element_data.pop(key)
+    return element_data
+
+
+def find_existing_node(graph: MultiDiGraph, new_node: Point) -> Union[Tuple[int, dict], Tuple[None, None]]:
```
Comment (Carsopre) [inline ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py:229]: Docstrings please.

File: ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py [hunk 6]
```diff
@@ -93,3 +96,151 @@ def from_shapefile_to_poly(shapefile: Path, out_path: Path, outname: str = ""):
...
+    for node, data in graph.nodes(data=True):
+        # Assuming the x and y coordinates are stored in data as 'x' and 'y'
+        if not data.get('geometry', ''):
+            raise ValueError("Nodes should have geometry")
+
+        if data['geometry'] == new_node:
+            return node, data
+    return None, None
+
+
```
Comment (Carsopre) [inline ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py:239]: Docstrings please.

**Review by Carsopre — CHANGES_REQUESTED** [review]
@sahand-asgarpour at this moment I see too many functions which are a copy & paste (or variations) from the `osmnx` library. I believe it is a better approach to just have a `osmnx_custom_library.py` file with such methods so that it is more traceable. Keep in mind that I do not support this approach as it can bring us more dependencies as it was demonstrated on #410 .

On the other hand, you should be able to start writing small (pytests) tests to verify the functionalities being written.

**Comment by sahand-asgarpour** [comment]
> @sahand-asgarpour at this moment I see too many functions which are a copy & paste (or variations) from the `osmnx` library. I believe it is a better approach to just have a `osmnx_custom_library.py` file with such methods so that it is more traceable. Keep in mind that I do not support this approach as it can bring us more dependencies as it was demonstrated on #410 .
> 
> On the other hand, you should be able to start writing small (pytests) tests to verify the functionalities being written.

**Review by Carsopre — APPROVED** [review]

**Pull request merged**